core/shared/src/main/scala/scalaz/zio/delay/Delay.scala

```diff
@@ -0,0 +1,81 @@
+package scalaz.zio.delay
+
+import java.util.concurrent.TimeUnit
+
+import scalaz.zio.ZIO
+import scalaz.zio.clock.{ Clock, _ }
+import scalaz.zio.delay.Delay._
+import scalaz.zio.duration.Duration
+
+/**
+ * Delay represents how much time an effect must wait before start.
+ * Delays can be relative to current time (a given duration + current time) or an absolute point in time
+ * Delays make waiting less cpu intensive and avoids clock drifts caused by context switching or hibernation
+ */
+sealed trait Delay extends Serializable with Product { self =>
```

Comment:
`sealed trait` should be replaced by `sealed abstract class`.

Reply:
Just for curiosity sake: what's the difference between them? I still don't get some optimizations related to traits/abstract classes

Reply:
I believe the reason is the bytecode generated is more efficient, due to how methods are ordered in the virtual table. With abstract classes, the order is predictable and consistent, which can be optimised (I think this leads to `invokevirtual` calls being inlined, but don't quote me on this 😅); that is not the case with traits, and the bytecode (`invokeinterface`) cannot be optimised.

Reply:
Here's another reason:

<https://typelevel.org/blog/2018/05/09/product-with-serializable.html>

Speaking of which we should test extending `Any` with the modules...